### PR TITLE
External computed by

### LINKED ISSUE #2: Introduce computed_by fields, with external plugins
Right now, you can define worksheets, and operate on them [[1](https://github.com/helloeave/worksheets/blob/master/worksheets_test.go#L21)]. These can also be saved, and loaded from the db [[2](https://github.com/helloeave/worksheets/blob/master/dbstore_test.go#L23)]. And there is also a convenient testing framework [[3](https://github.com/helloeave/worksheets/blob/master/features/example.feature)].

What the next set of key functionality is, revolves around editing. To avoid building a full blown programming language, my idea is to have code written in Golang which can compute derived values. So if we have a field `first_name` and a field `last_name`, and that we want to have a derived value `full_name`, we'd write the concat code in Golang. This would look like:

```
worksheet simple {
  1:first_name text
  2:last_name text
  3:full_name text computed_by { external }
  4: age number[0]
}
```

However, to inscribe itself in the framework, we want to make sure that the inputs to `full_name` are precisely captured, so that we do not recompute `full_name` if another unrelated field (e.g `age`) gets updated.

Initial idea to express this would be to have an interface `ComputedBy`  with a method `Args() []string` describing the fields needed, e.g.` ["first_name", "last_name"]`, and a method `Compute(args ...Value) Value`which takes the extracted values according to the args specified, and produces a correct value.

The hope with this is that when we do write the language (or write parts of it), we can simply plug in the runtime into this structure, and have the rest of the framework untouched.

### PR DESCRIPTION
**Pull request by alexandrinaw**
Fixes #2 😀 

i'm sure there's lots to be cleaned up but some 👀  now would be much appreciated!

### CODE REVIEW

File: worksheets_test.go [hunk 1]
```diff
@@ -84,6 +86,227 @@ func (s *Zuite) TestWorksheetNew_multipleDefsSameName() {
 	}
 }
 
+func (s *Zuite) TestExample_externalComputedBy() {
```
Comment (pascallouisperez) [inline worksheets_test.go]: Let's turn all these 'negative' tests into
```
cases := []struct{
  def string
  opt *Options
  expectedErr string
}{
  ...
}
for _, ex := range cases {
  var opts []Options
  if ex.opt != nil {
    opts = append(opts, opt)
  }
  err := NewDefinitions(strings.NewReader(ex.def), ...opts)
  ... etc.
}
```

applicable for `TestExample_externalComputedBy` to `TestExample_externalComputedBy4`

File: worksheets_test.go [hunk 2]
```diff
@@ -84,6 +86,227 @@ func (s *Zuite) TestWorksheetNew_multipleDefsSameName() {
...
+type fullName []string
+
+var _ ComputedBy = fullName([]string{})
+
+func (fn fullName) Args() []string {
+	return fn
+}
+
+func (fn fullName) Compute(values ...Value) Value {
+	firstName := values[0].String()
```
Comment (pascallouisperez) [inline worksheets_test.go]: Can you add a `// TODO(pascal): we need to figure out how to make Values useful, e.g. having an AsString()`

File: worksheets_test.go [hunk 3]
```diff
@@ -84,6 +86,227 @@ func (s *Zuite) TestWorksheetNew_multipleDefsSameName() {
...
+	require.NoError(s.T(), err)
+
+	ws := defs.MustNewWorksheet("complicated")
+
+	require.False(s.T(), ws.MustIsSet("full_name"))
+
+	ws.MustSet("first_name", NewText("Alice"))
+	ws.MustSet("last_name", NewText("Maters"))
+	ws.MustSet("birth_year", MustNewValue("1945"))
+	require.Equal(s.T(), `"Alice Maters, age 73, born in 1945"`, ws.MustGet("bio").String())
```
Comment (pascallouisperez) [inline worksheets_test.go]: Sweet!

Can you also add assertions on `full_name` and on `age` before the full `bio`?

File: worksheets_test.go [hunk 4]
```diff
@@ -84,6 +86,227 @@ func (s *Zuite) TestWorksheetNew_multipleDefsSameName() {
...
+
+var _ ComputedBy = fullName([]string{})
+
+func (fn fullName) Args() []string {
+	return fn
+}
+
+func (fn fullName) Compute(values ...Value) Value {
+	firstName := values[0].String()
+	lastName := values[1].String()
```
Comment (pascallouisperez) [inline worksheets_test.go]: Thought about this more. Here you can cast to get the actual text `values[1].(*tText).value`.

(As we improve the library, we'll want to have a better API for `Value`, but that's ok for now.)

File: worksheets_test.go [hunk 5]
```diff
@@ -84,6 +86,227 @@ func (s *Zuite) TestWorksheetNew_multipleDefsSameName() {
...
+type age []string
+
+var _ ComputedBy = age([]string{})
+
+func (fn age) Args() []string {
+	return fn
+}
+
+func (fn age) Compute(values ...Value) Value {
+	birthYearStr := values[0].String()
```
Comment (pascallouisperez) [inline worksheets_test.go]: Here `birthYear := values[0].(*tNumber).value` will do the trick for now.

File: worksheets_test.go (hunk 4 above)
Comment (alexandrinaw) [reply worksheets_test.go]: Thanks! Is there anything nicer than a switch statement for the cases where one of the values hasn't been set yet and is therefore `*tUndefined`?

File: parser.go [hunk 6]
```diff
@@ -41,6 +41,10 @@ type tWorksheet struct {
 	fields        []*tField
 	fieldsByName  map[string]*tField
 	fieldsByIndex map[int]*tField
+
+	// derived values handling
+	externals  map[int]ComputedBy
+	dependents map[int][]int
```
[the remark this answers is not in the document]
Comment (alexandrinaw) [reply parser.go]: no worries, renaming is easy w/ find and replace!

**Comment by pascallouisperez** [comment]
LGTM! merge away

**Pull request merged**